_posts/2023-12-20-standard-libraries.markdown

```diff
@@ -0,0 +1,407 @@
+and not something like a `Wrappers.Option<nat>` or a `Wrappers.Result<nat, string>`,
+because it always succeeds in parsing the string.
+
+How does Dafny know that `resultStr` is always parsable as a non-negative integer?
+Because of the post-conditions of `Seq.IndexByOption` and `Seq.LastIndexByOption`,
+Dafny actually **knows** that they return the indexes of elements that satisfy the "by" predicate,
+and therefore deduces that `[line[firstDigitIndex], line[lastDigitIndex]]`
+is a string that only contains digits.
+I don't know about you but I think that's super cool.
+
+## Bits and Bytes
+
+So now that we've solved the most interesting bit of the puzzle,
+we just have to feed the actual puzzle input into our Dafny code.
+Let's assume we've downloaded the puzzle to `input.txt` in the current directory.
+How the heck do we read this with Dafny?
+
+Good news! There's a standard library for that too: `Std.FileIO`.
+This one is very basic for now, explicitly meant for simple cases of reading and writing
+file data for these kinds of use cases, rather than modelling entire file systems.
+Here's our first attempt at a reusable utility for reading Advent of Code puzzle input
+(because after all we have another month's worth of puzzles to solve!)
+
+```dafny
+// Just the imports we need for this snippet.
+import Std.FileIO
+import Std.Wrappers
+
+method ReadPuzzleInput() returns (input: Wrappers.Result<string, string>) {
+  input := FileIO.ReadBytesFromFile("input.txt");
+}
+```
+
+Note that `FileIO.ReadBytesFromFile` is a `method` rather than a `function`,
+and so `ReadPuzzleInput` has to be as well.
+A `function` has to behave like a pure mathematical function
+in order for Dafny to reason about the behavior of programs:
+it has to produce the same output every time it is given the same input.
+This is clearly not true for an operation to read the contents of a file,
+so `ReadBytesFromFile` has to be a method, which is allowed to behave non-deterministically.
+
+Again we have a typechecking error: we want to get the puzzle input as a string,
+but `ReadBytesFromFile` produces bytes, more specifically a `seq<bv8>`.
+`bv8` is short for "bit vector of length 8", equivalent to a byte.
+We need to convert the bytes to characters somehow.
+
+Realistically, we'd assume that the puzzle input is in ASCII and just convert every byte directly
+to its corresponding character.
+But let's pretend we're writing Real Code for the moment so I can show off the `Std.Unicode` library;
+hey you never know, later puzzles might have proper UTF8 content!
+
+```dafny
+import Std.BoundedInts
+import Std.FileIO
+import Std.Unicode.UnicodeStringsWithUnicodeChar
+import Std.Wrappers
+
+method ReadPuzzleInput() returns (input: Wrappers.Result<string, string>) {
+  var bytesAsBVs :- FileIO.ReadBytesFromFile("input.txt");
+  
+  var bytes := seq(|bytesAsBVs|, i requires 0 <= i < |bytesAsBVs| => bytesAsBVs[i] as BoundedInts.uint8);
```

Comment:
It's unfortunate we need to do this conversion. We ought to be able to just cast. Anyway.

Reply:
Agreed, also noted in https://github.com/dafny-lang/dafny/issues/4878